aten/src/THC/THCAtomics.cuh

```diff
@@ -12,36 +12,42 @@ struct AtomicAddIntegerImpl;
 template<typename T>
 struct AtomicAddIntegerImpl<T, 1> {
   inline __device__ void operator()(T *address, T val) {
-    uint32_t * address_as_ui =
-        (uint32_t *) (address - ((size_t)address & 3));
+    size_t offset = (size_t)address & 3;
+    uint32_t * address_as_ui = (uint32_t *)((char *)address - offset);
     uint32_t old = *address_as_ui;
-    uint32_t shift = (((size_t)address & 3) * 8);
+    uint32_t shift = offset * 8;
     uint32_t sum;
+    uint32_t newval;
     uint32_t assumed;
 
     do {
       assumed = old;
-      sum = val + T((old >> shift) & 0xff);
-      old = (old & ~(0x000000ff << shift)) | (sum << shift);
-      old = atomicCAS(address_as_ui, assumed, old);
+      sum = (old >> shift) & 0xff;
+      T tmp = THCNumerics<T>::add(val, sum);
+      sum = *reinterpret_cast<uint8_t*>(&tmp);
```

Comment:
Sorry, I lost my train of thought from last week -- why is this necessary again? What happens if we don't have this? An example would help here.

Reply:
This seems equivalent to `sum = THCNumerics<T>::add(val, sum)`

Reply:
if I do:
```
   int8_t a = -5;
   uint32_t b = *reinterpret_cast<uint32_t*>(&a);
   uint32_t c = (uint32_t)a;
```
b != c

if it was just the change in sign they'd be the same, but since it changes sign and size the interpreted value changes. 

as an example, -1 to unsigned is always ~0, so if the destination is bigger it fills in 1's in bits we didn't want to change. we just want the raw bytes of the source memory space.

Reply:
Got it, thank you for the example, that helped a lot.

The reinterpret_cast might violate strict aliasing and could lead to UB: see https://xania.org/200712/cpp-strict-aliasing and https://stackoverflow.com/questions/14623266/why-cant-i-reinterpret-cast-uint-to-int . Could we switch this to one of the solutions in the latter link?

Reply:
Also, could you include a comment in the code that summarizes what you've explained here?

Reply:
Are we sure the strict aliasing problem applies here? I'd rather not complicate the code if it's not actually necessary to add the structs it would take.

What I'm seeing is that if we have two pointers of different types, during optimization the compiler can assume that they are not aliases/pointers to the same location in memory when making optimizations.

as an example, if we had:
```
f(int* x, int * y) {
*x=0
*y=1
return *x;
```
`*y=1` _could_ overwrite the value in x, because the compiler doesn't know if y is a pointer to the same location as x. If x and y are pointers to different types, the return statement can be compiled as `return 0` because the compiler is expected to assume `*y=1` cannot change the return value (can't be an alias to x).

In this code, we're just copying the value out of a pointed to address and interpreting it differently. we don't maintain multiple pointers/aliases of different types (sum isn't a pointer), so I'm not seeing where we can actually get UB. There's a transient pointer, but it's just read from and thrown away. For the compiler to pretend that temporary address isn't an alias doesn't seem like it would be problematic.

Also it looks like we compile with -fno-strict-aliasing

Reply:
I'm not sure so I am raising the question. The link here: https://stackoverflow.com/questions/14623266/why-cant-i-reinterpret-cast-uint-to-int looks like exactly what we're doing (scroll down to the section on reinterpret_cast)

Reply:
I think I have a simpler change you'll like: if I initially cast to an unsigned type of the same size, there's no extra space to pad with 1's in the case of a negative signed value. This seems to work with a regular cast without needing to reinterpret bytes, use a union, or memcpy.

Reply:
Yeah, this is much better now, thank you for finding it